### PR TITLE
tsview: reconstruct and plot temporal model

### PR DESCRIPTION
**Pull request by yuankailiu**
**Description of proposed changes**

Hello! I would like to propose to strengthen the current `tsview.py` code to have more abilities on time-series investigation (https://github.com/nisar-solid/ATBD/issues/4).

## add `--show-model` arguments to allow for ts2vel model fitting:

By appending `--show-model` or `--model`, we can plot the ts2vel model fit (default: False) on the interactive figure.

By specifying `--poly`, `--periodic`, `--step`, `--exp`, `--log`, the code will call `timeseries2velocity.py` here to make a much complex temporal model. The default of the model, if not specified, is a linear function (degree-1 polynomial).

Check the `--help` for details

## reconstruct the predicted time series based on model parameters:

Using the above model parameters, reconstruct a predicted time series, plot it in the interactive figure (if `--show-model`).

## add `--savetxt` or `--txt` argument to save ts and model param to file:

By setting `--savetxt = custom_filename.txt`, we can save the model parameters and the observed and predicted time series to a text file. This could be useful for further customized plotting or utilization out of MintPy package.

## `timeseries2velocity.py`: raise an error when G matrix redundant:

The G matrix could be redundant (has another almost all-one column), or even rank deficient (with another all-one column) when the exp and log functions have very small characteristic times (tau), or when their onset times are way too early than the minimum date of the time-series data. Now the code will raise an error if this occurs to end the process and remind users to try a different tau or onset time.

## `ptime.py`: add a function to produce a range of dates between date_min and date_max:

This is just a function that been used in the updated `tsview.py` code to get a reconstructed time series with finer temporal resolution. By calling `date2arange(dmin, dmax)` (both are strings in YYYYMMDD format), it returns a list of date strings (in YYYYMMDD format) which has an interval of one day.

**Reminders**

- [x] Pass Codacy code review (green)
- [x] Pass Circle CI test (green)
- [x] Make sure that your code follows our style. Use the other functions/files as a basis.
- [x] If modifying functionality, describe changes to function behavior and arguments in a comment below the function declaration.
- [x] If adding new functionality, add a detailed description to the documentation and/or an example.

### CODE REVIEW

File: mintpy/tsview.py [hunk 1]
```diff
@@ -81,6 +84,21 @@ def create_parser():
...
+                      help='a polynomial function with the input degree (default: %(default)s).')
+    model.add_argument('--periodic', '--period', '--peri', dest='periodic', type=float, nargs='+', default=[],
+                      help='periodic function(s) with period in decimal years (default: %(default)s).')
+    model.add_argument('--step', dest='step', type=str, nargs='+', default=[],
+                      help='step function(s) at YYYYMMDD (default: %(default)s).')
+    model.add_argument('--exp', '--exponential', dest='exp', type=str, nargs='+', action='append', default=[],
+                      help='exponential function(s) at YYYYMMDD with characteristic time(s) tau in decimal days (default: %(default)s).')
+    model.add_argument('--log', '--logarithmic', dest='log', type=str, nargs='+', action='append', default=[],
+                      help='logarithmic function(s) at YYYYMMDD with characteristic time(s) tau in decimal days (default: %(default)s).')
+
```
Comment (yunjunz) [inline mintpy/tsview.py:101]: With this usage, it is a good time to move them into a func in arg_group.py and re-used it in both ts2vel.py and tsview.py.

Comment (yuankailiu) [reply mintpy/tsview.py:101]: Ok, I can do that.

File: mintpy/tsview.py [hunk 2]
```diff
@@ -111,6 +129,42 @@ def cmd_line_parse(iargs=None):
...
+                        inps.logDict[onset_time] = np.array(char_times).astype(float).tolist()
+                    else:
+                        msg = 'NO characteristic time found: {}\n'.format(char_times)
+                        msg += 'one or more characteristic time(s) are required for each onset date for the log function, e.g.:\n'
+                        msg += '--exp 20181026 60 OR\n'
+                        msg += '--exp 20161231 80.5 200  # append as many char_times as you like!'
+                        raise ValueError(msg)
+                else:
+                    raise ValueError('input onset time is NOT in YYYYMMDD format: {}'.format(onset_time))
+
```
Comment (yunjunz) [inline mintpy/tsview.py:167]: This part is duplicated from ts2vel.py. Could we move them, in ts2vel.py, to a function and call them here?

Comment (yuankailiu) [reply mintpy/tsview.py:156]: Ok, that makes sense. Thanks

**Comment by yunjunz** [comment]
Below are two examples for the new time func fitting plot with `tsview.py`, for the 2008 Wells EQ and Fernandina example datasets:
WellsTS
FernandinaTS

**Comment by yuankailiu** [comment]
These updates are great! Thanks, @yunjunz.

**Comment by yunjunz** [comment]
I am glad you like it. Now, `tsview.py` is really powerful and convenient!

**Comment by yuankailiu** [comment]
Is it easy to also make the output text file with the correct unit?

**Comment by yunjunz** [comment]
@yuankailiu the unit is handled in the text file, here is the one after running the cmd below:

```bash
tsview.py timeseries_SET_ERA5_ramp_demErr.h5 --yx 215 310 --poly 5 --step 20170910 --periodic 0.5 --save
```

```
# time-series file=['timeseries_SET_ERA5_ramp_demErr.h5']
# Y/X = 215, 310, lat/lon = -0.3676, -91.5344
# reference pixel: y=76, x=156
# reference date: 20180619
# estimated time function parameters:
#     velocity           :   -1.81 +/-    3.84 cm/year
#     acceleration       :    2.06 +/-   13.53 cm/year^2
#     poly3              :   39.88 +/-   30.00 cm/year^3
#     poly4              : -105.40 +/-   38.65 cm/year^4
#     poly5              :   94.05 +/-   22.26 cm/year^5
#     semiAnnualAmplitude:    0.44 cm
#     semiAnnualPhase    :    1.52 radian
#     step20170910       :   -5.72 +/-    0.74 cm
# unit: cm
20141213	-38.352062
20141225	-38.195
20150307	-38.038654
20150319	-38.026726
20150331	-37.256172
20150412	-36.184864
20150424	-36.84712
20150506	-36.85152
20150518	-37.101288
20150530	-35.759403
20150611	-36.648758
20150623	-37.5262
20150717	-37.40409
20150729	-37.110847
20150822	-35.617863
20150903	-36.09602
20150915	-35.758038
20150927	-36.712673
20151009	-36.1088
20151021	-35.98331
20151102	-38.01382
20151114	-35.574276
20151126	-35.83935
20151208	-36.034473
20151220	-35.297623
20160101	-35.916786
20160113	-35.659107
20160125	-36.08254
20160206	-34.39466
20160218	-33.459553
20160301	-34.07693
20160406	-33.767937
20160418	-32.510437
20160430	-33.821796
20160512	-32.232796
20160524	-32.205112
20160605	-32.242264
20160629	-29.748371
20160711	-31.29574
20160723	-30.92281
20160804	-29.928732
20160816	-27.765469
20160828	-28.022406
20160909	-29.233456
20160921	-29.69147
20161003	-27.38938
20161015	-29.386044
20161027	-28.259987
20161108	-27.86995
20161120	-29.557163
20161202	-28.168816
20161214	-28.848883
20161226	-29.15377
20170107	-28.640348
20170119	-27.805033
20170131	-26.522005
20170212	-27.127384
20170224	-29.348785
20170308	-28.699139
20170320	-29.816788
20170401	-29.900873
20170413	-27.57073
20170425	-27.81479
20170507	-27.146425
20170519	-27.700243
20170531	-28.638428
20170612	-25.683563
20170624	-26.169432
20170706	-27.37354
20170718	-27.077103
20170730	-25.374985
20170811	-25.84183
20170823	-24.817177
20170904	-21.511293
20170916	-30.718029
20170928	-29.699802
20171010	-28.155863
20171022	-28.21426
20171103	-27.5161
20171115	-27.449083
20171127	-26.581001
20171209	-25.586145
20171221	-25.318087
20180102	-23.876528
20180114	-23.311441
20180126	-21.016935
20180207	-19.251268
20180219	-17.573822
20180303	-15.904519
20180315	-14.485565
20180327	-13.031757
20180408	-10.6722145
20180420	-8.240679
20180502	-6.290036
20180514	-5.5675774
20180526	-4.2038174
20180607	-2.9387653
20180619	0.0
```

**Comment by yuankailiu** [comment]
Thanks. This is really better than the quick-and-dirty version that I did before.

**Comment by yunjunz** [comment]
I love the new tsview.py and am very happy we made this happen @yuankailiu!

**Review by yunjunz — APPROVED** [review]

**Pull request merged**